File: 0424_longestRepeatingCharacterReplacement/main.cpp
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        int leftPtr = 0;
        int rightPtr = 0;
        int charCount[26] = {};
        int res = 0;
        for(int rightPtr = 0; rightPtr < s.length(); rightPtr++)
        {
            charCount[s[rightPtr]-'A']++;
            if((rightPtr-leftPtr+1)- *max_element(charCount, charCount+26) <= k)
            {
                res = rightPtr - leftPtr + 1;
            }
            else
            {
                charCount[s[leftPtr]-'A']--;
                leftPtr++;
            }
        }
        return res;
    }

    int characterReplacement1(string s, int k) {
        int leftPtr = 0;
        int rightPtr = 0;
        int maxCount = 0;
        int charCount[26] = {};
        int res = 0;
        for(int rightPtr = 0; rightPtr < s.length(); rightPtr++)
        {
            charCount[s[rightPtr]-'A']++;
            maxCount = max(charCount[s[rightPtr]-'A'], maxCount);
            if((rightPtr-leftPtr+1)- maxCount <= k)
            {
                res = rightPtr - leftPtr + 1;
            }
            else
            {
                charCount[s[leftPtr]-'A']--;
                leftPtr++;
            }
        }
        return res;
    }
    
};

File: 0424_longestRepeatingCharacterReplacement/main.py
class Solution:
    # O(26n)
    def characterReplacement(self, s: str, k: int) -> int: 
        count = {}
        res = 0
        
        l = 0
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)

            while (r-l+1) - max(count.values()) > k:
                count[s[l]] -= 1
                l += 1
            res = max(res, r-l+1)
        return res
    
    # O(n)
    def characterReplacement1(self, s: str, k: int) -> int:
        count = {}
        res = 0
        maxFreq = 0
        l = 0
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)
            # we don't need to update maxFreq in time, cause we focus on "longest" substring which is "maxFreq +k"
            # If we can't surpass historic maxFreq, then we won't be able to generate a longer valid substring.
            maxFreq = max(maxFreq, count[s[r]])
            while (r-l+1) - maxFreq > k:
                count[s[l]] -= 1
                l += 1
            res = max(res, r-l+1)
        return res
